src/codegen/x86_64/arch_x86_64.c

```diff
@@ -380,8 +383,1257 @@ static enum IndirectJumpType negate_jump(enum IndirectJumpType j) {
+  switch (inst) {
+  default: ICE("ERROR: mcode_imm_to_mem(): Unsupported instruction %d (%s)", inst, instruction_mnemonic(context, inst));
+  }
+}
+
+static void mcode_mem_to_reg(CodegenContext *context, enum Instruction inst, RegisterDescriptor address_register, int64_t offset, RegisterDescriptor destination_register, enum RegSize size) {
+  switch (inst) {
+
+  case I_LEA: {
+
+    switch (size) {
+    case r8: ICE("x86_64 machine code backend: LEA does not have an 8-bit encoding.");
+    case r16: {
+      // 0x66 + 0x8d /r
+      uint8_t sixteen_bit_prefix = 0x66;
+      fwrite(&sixteen_bit_prefix, 1, 1, context->machine_code);
+    } // FALLTHROUGH to case r32
+    case r32: {
+      // 0x8d /r
+      uint8_t op = 0x8d;
+
+      // Encode a REX prefix if either of the ModRM register descriptors need
+      // the bit extension.
+      uint8_t address_regbits = regbits(address_register);
+      uint8_t destination_regbits = regbits(destination_register);
+      if (REGBITS_TOP(address_regbits) || REGBITS_TOP(destination_regbits)) {
+        uint8_t rex = rex_byte(false, REGBITS_TOP(destination_regbits), false, REGBITS_TOP(address_regbits));
+        fwrite(&rex, 1, 1, context->machine_code);
+      }
+
+      // Mod == 0b10  ->  (R/M)+disp32
+      // Reg == Destination
+      // R/M == Address
+      uint8_t modrm = modrm_byte(0b10, destination_regbits, address_regbits);
+
+      fwrite(&op, 1, 1, context->machine_code);
+      fwrite(&modrm, 1, 1, context->machine_code);
+      int32_t disp32 = (int32_t)offset;
+      fwrite(&disp32, 4, 1, context->machine_code);
+
+    } break;
+    case r64: {
+      // REX.W + 0x8d /r
+      uint8_t op = 0x8d;
+
+      // Encode a REX prefix if either of the ModRM register descriptors need
+      // the bit extension.
+      uint8_t address_regbits = regbits(address_register);
+      uint8_t destination_regbits = regbits(destination_register);
+      uint8_t rex = rex_byte(true, REGBITS_TOP(destination_regbits), false, REGBITS_TOP(address_regbits));
+
+      // Mod == 0b10  ->  (R/M)+disp32
+      // Reg == Destination
+      // R/M == Address
+      uint8_t modrm = modrm_byte(0b10, destination_regbits, address_regbits);
+
+      fwrite(&rex, 1, 1, context->machine_code);
+      fwrite(&op, 1, 1, context->machine_code);
+      fwrite(&modrm, 1, 1, context->machine_code);
+      int32_t disp32 = (int32_t)offset;
+      fwrite(&disp32, 4, 1, context->machine_code);
+
+    } break;
+    } // switch (size)
+
+  } break; // case I_LEA
+
+  case I_MOV: {
+
+    switch (size) {
+    case r8: {
+      // 0x8a /r
+      uint8_t op = 0x8a;
+
+      // Encode a REX prefix if either of the ModRM register descriptors need
+      // the bit extension.
+      uint8_t address_regbits = regbits(address_register);
+      uint8_t destination_regbits = regbits(destination_register);
+      uint8_t rex = rex_byte(true, REGBITS_TOP(address_regbits), false, REGBITS_TOP(destination_regbits));
```

Comment:
This looks like `r64` handling, but it's under `r8`. Maybe copy-and-paste error? I don't think there should *always* be a REX prefix for `r8/m8 to r8` opcode...